multiset_test.go

```diff
@@ -0,0 +1,315 @@
+		if i != removeIndex {
+			data, err := hex.DecodeString(test.dataElementHex)
+			if err != nil {
+				t.Fatal(err)
+			}
+			m2.Add(data)
+		}
+	}
+	m2.Remove(removeData)
+
+	if m1.Finalize() != m2.Finalize() {
+		t.Fatalf("m1 and m2 was exepcted to have the same hash, but got instead m1 %s and m2 %s", m1.Finalize(), m2.Finalize())
+	}
+}
+
+func TestParseMultiSetFail(t *testing.T) {
+	data := [64]byte{}
+	copy(data[:], Secp256k1Order.Bytes())
+	_, err := ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("Shouldn't be able to parse a multiset bigger with x bigger than the field size")
+	}
+	data = [64]byte{}
+	copy(data[32:], Secp256k1Order.Bytes())
+	_, err = ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("Shouldn't be able to parse a multiset bigger with y bigger than the field size")
+	}
+	set := NewMultiset()
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("Failed generating random data %v %d", err, n)
+	}
+	set.Add(data[:])
+
+}
+
+func TestMultiSet_Reset(t *testing.T) {
+	set := NewMultiset()
+	emptySet := NewMultiset()
+	data := [100]byte{}
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("Failed generating random data %v %d", err, n)
+	}
+	set.Add(data[:])
+	if set.Finalize() == emptySet.Finalize() {
+		t.Errorf("Set shouldn't be empty")
+	}
+	set.Reset()
+	if set.Finalize() != emptySet.Finalize() {
+		t.Errorf("Set should be empty")
+	}
+}
+
+func TestMultiSetAddRemove(t *testing.T) {
+	list := [LoopsN][100]byte{}
+	set := NewMultiset()
+	set2 := set
+	serializedEmpty := set.Serialize()
+	for i := 0; i < LoopsN; i++ {
+		data := [100]byte{}
+		n, err := rand.Read(data[:])
+		if err != nil || n != len(data) {
+			t.Fatalf("Failed generating random data %v %d", err, n)
+		}
+		set.Add(data[:])
+		list[i] = data
+	}
+	if set.Finalize() == set2.Finalize() {
+		t.Errorf("sets are the same when they should be different: set %x\n", set.Finalize())
+	}
+
+	for i := 0; i < LoopsN; i++ {
+		set.Remove(list[i][:])
+	}
+	if set.Finalize() != set2.Finalize() {
+		t.Errorf("sets are different when they should be the same: set1: %x, set2: %x\n", set.Finalize(), set2.Finalize())
+	}
+	if set.Serialize() != serializedEmpty {
+		t.Errorf("serialized sets are different when they should be the same: set1: %x, set2: %x\n", set.Serialize(), serializedEmpty)
+	}
+	parsedSet, err := ParseMultiSet(serializedEmpty)
+	if err != nil {
+		t.Errorf("error: '%v' happened when parsing: %v", err, serializedEmpty)
+	} else if parsedSet.Finalize() != set.Finalize() {
+		t.Errorf("sets are different when they should be the same: set1: %x, parsedSet: %x\n", set.Finalize(), parsedSet.Finalize())
+	}
+}
+
+func BenchmarkMultiSet_Add(b *testing.B) {
+	b.ReportAllocs()
+	list := make([][100]byte, b.N)
+	for i := 0; i < b.N; i++ {
+		data := [100]byte{}
+		n, err := rand.Read(data[:])
+		if err != nil || n != len(data) {
+			b.Fatalf("Failed generating random data %v %d", err, n)
+		}
+		list[i] = data
```

Comment:
I thought it might be able to move the array instead of copying it. but I guess not because it's all preallocated heap areas. you're right, I guess it's better to have slices and/or pointers in these cases. (although I know that in other languages copy can sometimes be faster than creating a reference, but I don't know if that's true for Go)